filters/filter.go

```diff
@@ -28,23 +28,38 @@ import (
 )
 
 var (
-	errInvalidFilterPattern        = errors.New("invalid filter pattern defined")
-	allowAllFilter          Filter = allowFilter{}
+	errInvalidFilterPattern                  = errors.New("invalid filter pattern defined")
+	allowAllFilter               Filter      = allowFilter{}
+	singleAnyCharFilterForwards  chainFilter = &singleAnyCharFilter{backwards: false}
+	singleAnyCharFilterBackwards chainFilter = &singleAnyCharFilter{backwards: true}
 )
 
 // LogicalOp is a logical operator
 type LogicalOp string
 
+// chainSegment is the part of the pattern that the chain represents
+type chainSegment int
+
 // A list of supported logical operators
 const (
 	// Conjunction is logical AND
 	Conjunction LogicalOp = "&&"
 	// Disjunction is logical OR
 	Disjunction LogicalOp = "||"
 
-	wildcardChar   = "*"
-	negationChar   = '!'
-	allowFilterStr = "All"
+	middle chainSegment = iota
+	start
+	end
+
+	allowFilterStr       = "All"
+	anyCharStr           = "AnyChar"
+	wildcardChar         = "*"
+	negationChar         = '!'
+	singleAnyChar        = '?'
+	singleRangeStartChar = '['
+	singleRangeEndStr    = "]"
+	rangeChar            = '-'
+	invalidNestedChars   = "?["
```

Comment:
Didn't want to allow for ranges inside the contains since it will result in multiple passes. Didn't put a check in there for all possible range chars, figured we can validate that stuff outside of this.